pkg/auth/serviceaccount.go

```diff
@@ -101,6 +101,25 @@ func EnsureOIDCServiceAccountExistsForResource(ctx context.Context, serviceAccou
 	return nil
 }
 
+// DeleteOIDCServiceAccountIfExists makes sure the given resource does not have an OIDC service account.
+// If it does that service account is deleted.
+func DeleteOIDCServiceAccountIfExists(ctx context.Context, serviceAccountLister corev1listers.ServiceAccountLister, kubeclient kubernetes.Interface, gvk schema.GroupVersionKind, objectMeta metav1.ObjectMeta) error {
+	saName := GetOIDCServiceAccountNameForResource(gvk, objectMeta)
+	sa, err := serviceAccountLister.ServiceAccounts(objectMeta.Namespace).Get(saName)
+
+	// Service Account Found
+	if err == nil && metav1.IsControlledBy(&sa.ObjectMeta, &objectMeta) {
+		logging.FromContext(ctx).Debugw("OIDC Service account already exists. Deleting OIDC service account")
+
+		err = kubeclient.CoreV1().ServiceAccounts(objectMeta.Namespace).Delete(ctx, sa.Name, metav1.DeleteOptions{})
+		if err != nil {
+			return fmt.Errorf("could not delete OIDC service account %s/%s for %s: %w", objectMeta.Name, objectMeta.Namespace, gvk.Kind, err)
+		}
+	}
```

Comment:
I think here you are missing returning an error (in the path where `err != nil`)
```suggestion
	} else if err != nil {
	    return err
	}
```

Alternatively, you could just `return err` at the end instead of `return nil`, but that might be less clear in terms of the intent of the code

Reply:
Actually, the only err that will be returned is Not Found error, which we would ignore, and the other error would be by GetByKey func which always returns the error as nil.
But it will be better to check if there are any changes in future so we can do - 
```
	} else if apierrs.IsNotFound(err) {
		return nil
	}

	return err
```